File: src/FSharpAux/KdTree.fs
namespace FSharpAux

/// Module for k -dimension trees  
module KdTree =

    /// k -dimension tree    
    type KdTree<'dataRecord,'coordinates> =
        | Node of KdTree<'dataRecord,'coordinates> * KdTree<'dataRecord,'coordinates> * 'coordinates  * int * 'dataRecord 
        | Leaf 




    /// Create KD Tree from list
    // converter -> converts 'dataRecord to dimension array ('coordinates)
    // getCoordinateByPos -> returns coordinate at position
    // k -> number of dimension
    let ofArrayBy (converter:'dataRecord -> 'coordinates) (getCoordinateByPos:'coordinates -> int -> 'coordinate) k arr =
        let rec loop data d k =
            match Array.isEmpty data with
            | true  -> Leaf
            | false ->            
                let axis = d % k
                Array.sortInPlaceBy (fun item -> (getCoordinateByPos (converter item) axis)) data
                let halfIndex = Array.length data / 2
                if halfIndex = 0 then
                    let dataPoint = data.[halfIndex]
                    Node (Leaf, Leaf, (converter dataPoint), d, dataPoint)
                else
                    let dataPoint = data.[halfIndex]
                    let lf, r =     data.[0..halfIndex - 1], data.[halfIndex+1..]
                    Node(loop lf  (d + 1) k, loop r (d + 1) k, (converter dataPoint), d, dataPoint)
        //let arr = data |> Array.ofSeq 
        loop arr 0 k



    /// Flattens KdTree
    let rec toList (tree:KdTree<'a,'b>) =
        match tree with
        | Leaf        -> []
        | Node(left, right, _, _, value) -> toList left  @ [value] @ toList right 
        

    /// Prints kd tree
    let print (kdTree:KdTree<'a,'b>) =    
            let rec printLoop (c) = seq {    
                match c with
                | Node (l,r,s,d,dt)        ->   yield  sprintf "\"Depth\" : %i \"Position\" : %A \"Data\" : %A, \n" d s dt 
                                                yield! printLoop l                                            
                                                yield! printLoop r                                             
                | Leaf                     ->   yield  sprintf ""
            }
            printLoop kdTree


    type Distance<'a> = 'a -> 'a -> float


    /// Searches query coordinates within a given radius
    let rangeSearch (distance:Distance<'coordinates>) (getCoordinateByPos:'coordinates -> int -> 'coordinate) (tree:KdTree<'dataRecord,'coordinates>) (radius) (queryCoordinates:'coordinates)  =
        let rec loop tree (acc:'a list) =
            match tree with 
            | Leaf -> acc
            | Node(cLeft, cRight, cCoor, cK, cValue) ->
                let d = distance cCoor queryCoordinates
                let acc' = if (d <= radius) then (cValue::acc) else acc
                let axis  = getCoordinateByPos queryCoordinates cK
                let cAxis = getCoordinateByPos cCoor cK
                match compare axis cAxis with
                | c when c < 0 -> 
                    if (abs (axis - cAxis) <= radius) then
                        loop cLeft (loop cRight acc')
                    
                    else
                        loop cLeft acc'
                
                | _ -> 
                    if (abs (axis - cAxis) <= radius) then
                        loop cRight (loop cLeft acc')
                    else
                        loop cRight acc'
        loop tree [] 



//let a = [|[|2.;1.;3.;|]; [|2.5;1.5;3.5;|]; [|1.;2.;1.;|]; [|12.;5.;2.4;|]; [|6.;1.;3.;|]; [|0.5;0.5;1.;|]|]
//
//
//let toCoordinates item = item
//let toPosition (item:float array) n = item.[n]
//let kd = ofArrayBy toCoordinates toPosition 3 a

//let euclidean v1 v2 = 
//    Seq.zip v1 v2
//    |> Seq.fold (fun d (e1,e2) -> d + ((e1 - e2) * (e1 - e2))) 0.
//    |> sqrt
//
//let test = rangeSearch euclidean toPosition kd 3. [|0.2; 0.2; 0.9|] 

File: tests/FSharpAux.Mocha.Tests/fable/src/FSharpAux.Core/KdTree.js
import { Union } from "../../fable_modules/fable-library.4.0.6/Types.js";
import { union_type, int32_type } from "../../fable_modules/fable-library.4.0.6/Reflection.js";
import { sortInPlaceBy } from "../../fable_modules/fable-library.4.0.6/Array.js";
import { comparePrimitives, compare } from "../../fable_modules/fable-library.4.0.6/Util.js";
import { cons, empty, singleton, append } from "../../fable_modules/fable-library.4.0.6/List.js";
import { append as append_1, singleton as singleton_1, delay } from "../../fable_modules/fable-library.4.0.6/Seq.js";
import { printf, toText } from "../../fable_modules/fable-library.4.0.6/String.js";

export class KdTree$2 extends Union {
    constructor(tag, fields) {
        super();
        this.tag = tag;
        this.fields = fields;
    }
    cases() {
        return ["Node", "Leaf"];
    }
}

export function KdTree$2$reflection(gen0, gen1) {
    return union_type("FSharpAux.KdTree.KdTree`2", [gen0, gen1], KdTree$2, () => [[["Item1", KdTree$2$reflection(gen0, gen1)], ["Item2", KdTree$2$reflection(gen0, gen1)], ["Item3", gen1], ["Item4", int32_type], ["Item5", gen0]], []]);
}

export function ofArrayBy(converter, getCoordinateByPos, k, arr) {
    const loop = (data, d, k_1) => {
        if (data.length === 0) {
            return new KdTree$2(1, []);
        }
        else {
            const axis = (d % k_1) | 0;
            sortInPlaceBy((item) => getCoordinateByPos(converter(item), axis), data, {
                Compare: compare,
            });
            const halfIndex = ~~(data.length / 2) | 0;
            if (halfIndex === 0) {
                const dataPoint = data[halfIndex];
                return new KdTree$2(0, [new KdTree$2(1, []), new KdTree$2(1, []), converter(dataPoint), d, dataPoint]);
            }
            else {
                const dataPoint_1 = data[halfIndex];
                const matchValue_1 = data.slice(0, (halfIndex - 1) + 1);
                const matchValue_2 = data.slice(halfIndex + 1, data.length);
                return new KdTree$2(0, [loop(matchValue_1, d + 1, k_1), loop(matchValue_2, d + 1, k_1), converter(dataPoint_1), d, dataPoint_1]);
            }
        }
    };
    return loop(arr, 0, k);
}

export function toList(tree) {
    if (tree.tag === 0) {
        return append(toList(tree.fields[0]), append(singleton(tree.fields[4]), toList(tree.fields[1])));
    }
    else {
        return empty();
    }
}

export function print(kdTree) {
    const printLoop = (c) => delay(() => ((c.tag === 1) ? singleton_1(toText(printf(""))) : append_1(singleton_1(toText(printf("\"Depth\" : %i \"Position\" : %A \"Data\" : %A, \n"))(c.fields[3])(c.fields[2])(c.fields[4])), delay(() => append_1(printLoop(c.fields[0]), delay(() => printLoop(c.fields[1])))))));
    return printLoop(kdTree);
}

export function rangeSearch(distance, getCoordinateByPos, tree, radius, queryCoordinates) {
    const loop = (tree_1_mut, acc_mut) => {
        loop:
        while (true) {
            const tree_1 = tree_1_mut, acc = acc_mut;
            if (tree_1.tag === 0) {
                const cRight = tree_1.fields[1];
                const cLeft = tree_1.fields[0];
                const cK = tree_1.fields[3] | 0;
                const cCoor = tree_1.fields[2];
                const acc$0027 = (distance(cCoor, queryCoordinates) <= radius) ? cons(tree_1.fields[4], acc) : acc;
                const axis = getCoordinateByPos(queryCoordinates, cK);
                const cAxis = getCoordinateByPos(cCoor, cK);
                if (comparePrimitives(axis, cAxis) < 0) {
                    if (Math.abs(axis - cAxis) <= radius) {
                        tree_1_mut = cLeft;
                        acc_mut = loop(cRight, acc$0027);
                        continue loop;
                    }
                    else {
                        tree_1_mut = cLeft;
                        acc_mut = acc$0027;
                        continue loop;
                    }
                }
                else if (Math.abs(axis - cAxis) <= radius) {
                    tree_1_mut = cRight;
                    acc_mut = loop(cLeft, acc$0027);
                    continue loop;
                }
                else {
                    tree_1_mut = cRight;
                    acc_mut = acc$0027;
                    continue loop;
                }
            }
            else {
                return acc;
            }
            break;
        }
    };
    return loop(tree, empty());
}

